### PR TITLE
First "real" implementation

### PR DESCRIPTION
**Pull request by juergenhemelt**
.. with sync of Schemas between Schema Registry and Egeria incl. mapping of schemas to KafkaTopics in Egeria.

### CODE REVIEW

**Comment by davidradl** [comment]
@juergenhemelt thanks for submitting. It looks like sonatype has identified 4 things to change.I will look through the code and review it more thoroughly today

File: README.md [hunk 1]
```diff
@@ -1,12 +1,46 @@
 <!-- SPDX-License-Identifier: CC-BY-4.0 -->
-<!-- Copyright Contributors to the Egeria project. -->
+<!-- Copyright Contributors to the ODPi Egeria project. -->
 
-# Egeria
+[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3044/badge)](https://bestpractices.coreinfrastructure.org/projects/3044)
 
-<!-- TODO: Add a README -->
```
Comment (davidradl) [inline README.md:6]: I suggest removing

File: README.md [hunk 2]
```diff
@@ -1,12 +1,46 @@
...
+[//]: # ([![Azure]&#40;https://dev.azure.com/odpi/egeria/_apis/build/status/odpi.egeria&#41;]&#40;https://dev.azure.com/odpi/Egeria/_build&#41;)
```
Comment (davidradl) [inline README.md:6]: I am not sure where you got these 3 lines from - could you direct me to any documentation that mentiones them please - as I have not added them to my readmes.

File: README.md [hunk 3]
```diff
@@ -1,12 +1,46 @@
...
+[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=odpi_egeria&metric=alert_status)](https://sonarcloud.io/dashboard?id=odpi_egeria)
+[![Maven Central](https://img.shields.io/maven-central/v/org.odpi.egeria/egeria)](https://mvnrepository.com/artifact/org.odpi.egeria)
 
-See https://egeria-project.org for the main Egeria Documentation.
 
-This is a new repository that we are currently setting up. 
+# Egeria integration connector for Event Schemas
```
Comment (davidradl) [inline README.md:11]: is this specific to Confluent Schema registry - if so then it should be in the connector name. I suspect we could have other event schema connectors for other technologies in the future

File: build.gradle [hunk 4]
```diff
@@ -34,11 +34,12 @@ description = 'New code for Egeria'
 
 // TODO: Specify Dependency versions as variables
 ext {
-    egeriaversion = '3.6'
+    egeriaversion = '3.10'
```
Comment (davidradl) [inline build.gradle:37]: I suggest using 3.12 the latest Egeria

File: src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import java.security.KeyManagementException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * This class implements the connection to a real Atlas System
```
Comment (davidradl) [inline src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java:27]: not Atlas.

File: src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+     *
+     * @return List of all subjects, if no subjects are listed, an empty array is returned.
+     */
+    @Override
+    public List<String> listAllSubjects() {
+        RestTemplate restTemplate;
+        try {
+            restTemplate = restTemplate();
+        } catch (Exception e) {
+            //TODO: Exception handling
```
Comment (davidradl) [inline src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java:53]: maybe create an issue for this and the other exception processing  to track separately

File: src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        return convertResponseToStringArray(jsonResponse);
+    }
+
+    /**
+     * REST call to retrieve the versions of a subject.
+     *
+     * @return List of all subjects, if no subjects are listed, an empty array is returned.
+     */
+    @Override
+    public String getSchema(String subject, String version) {
```
Comment (davidradl) [inline src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java:94]: maybe change version to schemaVersion and javadoc

File: src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+        String jsonResponse = responseEntity.getBody();
+
+        return convertResponseToStringArray(jsonResponse);
+    }
+
+    /**
+     * REST call to retrieve the versions of a subject.
+     *
+     * @return List of all subjects, if no subjects are listed, an empty array is returned.
```
Comment (davidradl) [inline src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/connection/ConfluentRestCalls.java:91]: this looks incorrect I think it is returing a schema.

File: build.gradle [hunk 9]
```diff
@@ -34,11 +34,12 @@ description = 'New code for Egeria'
...
     slf4jVersion = '1.7.36'
     jacksonVersion = '2.13.2'
     jupiterVersion = '5.8.2'
     httpclientVersion = '4.5.13'
+    mockitoVersion ='4.5.1'
```
Comment (davidradl) [inline build.gradle:42]: I think we need a settings.gradle file to supply the root project name. Unless there is some other way to do this.

File: src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/EventSchemaIntegrationProvider.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * EventSchemaIntegrationProvider is the connector provider for the kafka integration connector that extracts topic names from the broker.
+ */
+public class EventSchemaIntegrationProvider extends ConnectorProviderBase {
+
+    public static final String EVENT_SCHEMA_USER_ID = "event_schema_user_id";
+    public static final String EVENT_SCHEMA_PASSWORD = "event_schema_user_password";
+    private static final String connectorTypeGUID = "d64cbdb1-69bf-4da1-93cc-a8ba99b41428";
+    private static final String connectorTypeQualifiedName = "Event Schema Integration Connector";
```
Comment (davidradl) [inline src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/EventSchemaIntegrationProvider.java:21]: I think the name of this connector should include Confluent.

File: build.gradle [hunk 11]
```diff
@@ -34,11 +34,12 @@ description = 'New code for Egeria'
```
Comment (davidradl) [inline build.gradle:34]: the ext.name and description above should refer to this connector

File: README.md [hunk 12]
```diff
@@ -1,12 +1,46 @@
 <!-- SPDX-License-Identifier: CC-BY-4.0 -->
-<!-- Copyright Contributors to the Egeria project. -->
+<!-- Copyright Contributors to the ODPi Egeria project. -->
```
Comment (planetf1) [inline README.md:2]: this is the old label - we dropped ODPi in newer statements

File: README.md (hunk 2 above)
Comment (planetf1) [inline README.md:6]: we don't use azure builds any more, just github

File: README.md [hunk 13]
```diff
@@ -1,12 +1,46 @@
...
 
-# Egeria
+[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3044/badge)](https://bestpractices.coreinfrastructure.org/projects/3044)
 
-<!-- TODO: Add a README -->
+[//]: # ([![Azure]&#40;https://dev.azure.com/odpi/egeria/_apis/build/status/odpi.egeria&#41;]&#40;https://dev.azure.com/odpi/Egeria/_build&#41;)
+[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=odpi_egeria&metric=alert_status)](https://sonarcloud.io/dashboard?id=odpi_egeria)
```
Comment (planetf1) [inline README.md:7]: we're not using Sonar fully either - and not for this repo

File: README.md [hunk 14]
```diff
@@ -1,12 +1,46 @@
...
+[![Maven Central](https://img.shields.io/maven-central/v/org.odpi.egeria/egeria)](https://mvnrepository.com/artifact/org.odpi.egeria)
```
Comment (planetf1) [inline README.md:8]: If this repo produces a single artifact, you could point it specifically to the correct name - this above though is core egeria.

File: README.md [hunk 15]
```diff
@@ -1,12 +1,46 @@
...
+
+You will need to have an OMAG platform with the connector jar, a metadata server defined.
+
+### Testing mutations
+
+[//]: # (TODO: extend with testing information)
+
+----
+License: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/),
+Copyright Contributors to the ODPi Egeria project.
```
Comment (planetf1) [inline README.md:46]: I think we should drop the ODPi here too

File: build.gradle [hunk 16]
```diff
@@ -50,11 +51,14 @@ dependencies {
 
     // Needed for build and run
     implementation "org.slf4j:slf4j-api:${slf4jVersion}"
+    implementation "org.apache.httpcomponents:httpclient:${httpclientVersion}"
+    implementation "org.odpi.egeria:topic-integrator-api:${egeriaversion}"
```
Comment (planetf1) [inline build.gradle:55]: this should be 'compileOnly' - for all of the components we think are already in the chassis. It's more specific, and would avoid the libraries being picked up to include in any uber jar

File: build.gradle (hunk 11 above)
Comment (planetf1) [reply build.gradle:34]: it's worth checking the generated manifest for the maven artifact. The group is specified in the top level gradle. Names can default to directory name, but as you mention settings etc can override. Worth a second iteration once this is merged

File: build.gradle [hunk 17]
```diff
@@ -130,16 +134,15 @@ publishing {
...
                     }
                 }
                 scm {
-                    // TODO: Update references to git source for this particular repository
-                    connection = 'scm:git:git://github.com/odpi/egeria-template-newrepo.git'
-                    developerConnection = 'scm:git:ssh://github.com/odpi/egeria/egeria-template-newrepo.git'
-                    url = 'http://github.com/odpi/egeria-template-newrepo/'
+                    connection = 'scm:git:git://github.com/odpi/egeria-connection-integration-event-schema.git'
+                    developerConnection = 'scm:git:ssh://github.com/odpi/egeria-connection-integration-event-schema.git'
+                    url = 'http://github.com/odpi/egeria-connection-integration-event-schema/'
```
Comment (planetf1) [inline build.gradle:145]: use https on the web url Also this URL is incorrect (connection->connector)

File: src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/EventSchemaIntegrationConnector.java [hunk 18]
```diff
@@ -0,0 +1,187 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright © 2021 Atruvia AG <opensource@atruvia.de> as contributor to the ODPi Egeria project. */
```
Comment (planetf1) [inline src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/EventSchemaIntegrationConnector.java:2]: We should check our contribution rules for the project. We've used a common odpi contributors copyright with all code going into open source so far.

We need to check this before merging in .

Comment (planetf1) [reply src/main/java/org/odpi/openmetadata/adapters/connectors/integration/eventschema/EventSchemaIntegrationConnector.java:2]: This occurs in other files too - but I won't repeat.

**Review by planetf1 — CHANGES_REQUESTED** [review]

**Comment by planetf1** [comment]
I noticed lift had a few findings
 * there are some printStackTrace() calls, which are generally regarded as potentially insecure - they will likely show up as issues when we run code scans - however I see you have TODO tags, so presume this is known, and will be looked at later. Given they will cause scan warnings I think it may be worth opening up an issue to ensure the change isn't forgotten
 * There were a few unsynchronized method warnings around the connector management. Not looked at code, but could be worth checking those closely

**Comment by planetf1** [comment]
Note - Lift will block the merge with some of it's findings. Expand out the 'commands' section, to add comments which will cause lift to mark the issue to be ignored.

However they mostly look like valid issues

**Pull request closed without merging**